src/go/configgenerator/filtergen/backend_auth.go

```diff
@@ -35,38 +39,85 @@ const (
 )
 
 type BackendAuthGenerator struct {
-	// audMap is the list of unique audiences in the config.
-	audMap map[string]bool
+	// UniqueAudiences is the list of unique JWT audiences in the config.
+	UniqueAudiences map[string]bool
+
+	// AudienceBySelector lists the JWT audience for each method.
+	AudienceBySelector map[string]string
+
+	IamURL                  string
+	MetadataURL             string
+	HttpRequestTimeout      time.Duration
+	DependencyErrorBehavior string
+	BackendAuthCredentials  *options.IAMCredentialsOptions
+
+	AccessToken *helpers.FilterAccessTokenConfiger
 }
 
-// NewBackendAuthGenerator creates the BackendAuthGenerator with cached config.
-func NewBackendAuthGenerator(serviceInfo *ci.ServiceInfo) *BackendAuthGenerator {
-	return &BackendAuthGenerator{
-		audMap: getUniqueAudiences(serviceInfo),
+// NewBackendAuthFilterGensFromOPConfig creates a BackendAuthGenerator from
+// OP service config + descriptor + ESPv2 options. It is a FilterGeneratorOPFactory.
+func NewBackendAuthFilterGensFromOPConfig(serviceConfig *servicepb.Service, opts options.ConfigGeneratorOptions, params FactoryParams) ([]FilterGenerator, error) {
+	audienceBySelector, uniqueAudiences, err := GetJWTAudiencesBySelectorFromOPConfig(serviceConfig, opts)
+	if err != nil {
+		return nil, err
 	}
+	if len(uniqueAudiences) == 0 {
+		glog.Info("Not adding backend auth filter gens because there are no audiences.")
+		return nil, nil
+	}
+
+	return []FilterGenerator{
+		&BackendAuthGenerator{
+			UniqueAudiences:         uniqueAudiences,
+			AudienceBySelector:      audienceBySelector,
+			IamURL:                  opts.IamURL,
+			MetadataURL:             opts.MetadataURL,
+			HttpRequestTimeout:      opts.HttpRequestTimeout,
+			DependencyErrorBehavior: opts.DependencyErrorBehavior,
+			BackendAuthCredentials:  opts.BackendAuthCredentials,
+			AccessToken:             helpers.NewFilterAccessTokenConfigerFromOPConfig(opts),
+		},
+	}, nil
 }
 
 func (g *BackendAuthGenerator) FilterName() string {
 	return BackendAuthFilterName
 }
 
-func (g *BackendAuthGenerator) IsEnabled() bool {
-	return len(g.audMap) > 0
+func (g *BackendAuthGenerator) MatchAudience(selector string) (string, bool) {
+	if audience, ok := g.AudienceBySelector[selector]; ok {
+		return audience, true
+	}
+
+	// Try matching CORS selector.
+	originalSelector, err := CORSSelectorToSelector(selector)
+	if err != nil {
+		// Ignore error, assume no route match.
```

Comment:
Nit: add a log for the error too

Reply:
Changed logic here to separate against "no route match" and error.